### DISCUSSION TITLE
Add the option to have property URI buttons conditionally appear, if the property is present in the note

### DISCUSSION THREAD

**Discussion by FeralFlora** [chrisgurney/obsidian-note-toolbar#76] [Ideas]
As is, property URI buttons simply do nothing, if the given property is not present in the note. I think it would be nice to have the ability to hide the button, when the associated property is not present in the note. This way, the button being present can also act as a signal to encourage its effective use in this circumstance.

**Comment by chrisgurney** [comment]
@FeralFlora Thanks for the suggestion!

In an effort to keep the toolbar performant at the moment, toolbars are reused across notes where possible, so they don't have to be re-rendered. As such, the URLs (and variables within them) are calculated at _click_ time, versus at _render_ time.

The latter approach _might_ introduce unwanted flickering (if so, perhaps an option is needed to enable that, if it's OK with the user). In the process I would also want to look at exactly how toolbars are cached, and see if I can update just the relevant buttons.

- Related: Depending on overall effort required, I had also thought about updating labels in the toolbar as well, depending on the value of variables, so this enhancement would tie with that one, if there's a desire for that.

It's worth exploring, I believe, but a potential good chunk of time. I'll have to think about how it might work!

**Reply by FeralFlora** [reply]
> As such, the URLs (and variables within them) are calculated at click time, versus at render time.

It sounds like we are talking about slightly different things. I'm not talking about what the buttons do, but whether they appear, although that also falls under re-rendering, of course. I just thought that since the plugin already checks for overriding property specified toolbars, then it wouldn't be too different to check whether a property is present to determine if a button should be displayed or not.

But regardless, I totally get wanting to get it right, and ensure that the user experience remains performant. I also saw the related point about dynamic toolbar labels floating around here, and it sounds useful! 👍

**Comment by chrisgurney** [comment]
@FeralFlora Try out [version 1.8.3](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.8.3): If a link with variables is resolved to an empty value, then the item will be hidden from the toolbar. Let me know how it works for you!

**Reply by chrisgurney** [reply]
**Update:** I've noticed a bug(?) I'm taking a look at: If there is **no** prop in the note at all and the variable references it, I think the correct behaviour in this case is to: 
- hide the item if it's used in the URI,
- empty any value that references it (label, tooltip)

**Reply by chrisgurney** [reply]
**Update 2:** This should now be fixed with `1.8.4`.

**Reply by FeralFlora** [reply]
Nice, it seems to work, thank you!

**Reply by chrisgurney** [reply]
**Update 3:** I found another bug and have added support for menus (mobile only for now, but menus will be coming to toolbar items in the next release). 

Let me know if you run into any other issues!

**Reply by FeralFlora** [reply]
> **Update 3:** I found another bug and have added support for menus (mobile only for now, but menus will be coming to toolbar item in the next release).
> 
> Let me know if you run into any other issues!

Thanks, I will :)
